webpush_channels/views/subscriptions.py

```diff
@@ -13,8 +16,30 @@ class SubscriptionSchema(ResourceSchema):
     keys = KeySchema()
 
 
+class SHAendpoint(generators.Generator):
+
+    regexp = r'^[a-zA-Z0-9][a-zA-Z0-9_-]*$'
+
+    def __init__(self, request, config=None):
+        self.config = config
+        self._regexp = None
+        self.request = request.request
+
+        if not self.match(self()):
+            error_msg = "Generated record id does comply with regexp."
+            raise ValueError(error_msg)
+
+    def __call__(self):
+        return hashlib.sha224(json.dumps(self.request.json['data']['endpoint'])).hexdigest()
+
+
 @register(name='subscription',
           collection_path='/subscriptions',
           record_path='/subscriptions/{{id}}')
 class Subscription(UserResource):
     schema = SubscriptionSchema
+
+    def __init__(self, request, context=None):
+        super(Subscription, self).__init__(request, context)
+        if 'id' not in self.request.json['data']:
+            self.model.id_generator = SHAendpoint(self, self.request)
```

Comment:
You don't need to do that in a if, you should be able to just define model.id_generator = SHAendpoint